ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -115,8 +127,89 @@ public void start() throws ReplicationException {
+        Set<ReplicaId> removedPeerReplicas = removedReplicas.stream()
+            .filter(r -> partitionToPartitionInfo.containsKey(r.getPartitionId()) && r.getDataNodeId() != currentNode)
+            .collect(Collectors.toSet());
+
+        // No additional synchronization is required because cluster change handler of each dc only updates replica-threads
+        // belonging to certain dc. Hence, there is only one thread adding/removing remote replicas within a certain dc.
+
+        // 3. create replicaInfo for new remote replicas and assign them to replica-threads.
+        List<RemoteReplicaInfo> replicaInfosToAdd = new ArrayList<>();
+        for (ReplicaId remoteReplica : addedPeerReplicas) {
+          PartitionInfo partitionInfo = partitionToPartitionInfo.get(remoteReplica.getPartitionId());
+          // create findToken, remoteReplicaInfo
+          FindToken findToken =
+              this.tokenHelper.getFindTokenFactoryFromReplicaType(remoteReplica.getReplicaType()).getNewFindToken();
+          RemoteReplicaInfo remoteReplicaInfo =
+              new RemoteReplicaInfo(remoteReplica, partitionInfo.getLocalReplicaId(), partitionInfo.getStore(),
+                  findToken,
+                  TimeUnit.SECONDS.toMillis(storeConfig.storeDataFlushIntervalSeconds) * Replication_Delay_Multiplier,
+                  SystemTime.getInstance(), remoteReplica.getDataNodeId().getPortToConnectTo());
+          logger.info("Adding remote replica {} on {} to partition info.", remoteReplica.getReplicaPath(),
+              remoteReplica.getDataNodeId());
+          if (partitionInfo.addReplicaInfoIfAbsent(remoteReplicaInfo)) {
+            replicationMetrics.addMetricsForRemoteReplicaInfo(remoteReplicaInfo);
+          }
+          replicaInfosToAdd.add(remoteReplicaInfo);
+        }
+        addRemoteReplicaInfoToReplicaThread(replicaInfosToAdd, true);
+
+        // 4. remove replicaInfo from existing partitionInfo and replica-threads
+        List<RemoteReplicaInfo> replicaInfosToRemove = new ArrayList<>();
+        for (ReplicaId remoteReplica : removedPeerReplicas) {
+          PartitionInfo partitionInfo = partitionToPartitionInfo.get(remoteReplica.getPartitionId());
+          List<RemoteReplicaInfo> remoteReplicaInfos = new ArrayList<>(partitionInfo.getRemoteReplicaInfos());
+          for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
+            if (remoteReplicaInfo.getReplicaId().getDataNodeId() == remoteReplica.getDataNodeId()) {
+              logger.info("Removing remote replica {} on {} from replica threads.", remoteReplica.getReplicaPath(),
+                  remoteReplica.getDataNodeId());
+              replicaInfosToRemove.add(remoteReplicaInfo);
+              if (partitionInfo.removeRelicaInfoIfPresent(remoteReplicaInfo)) {
+                replicationMetrics.removeMetricsForRemoteReplicaInfo(remoteReplicaInfo);
+              }
+              break;
+            }
+          }
+        }
```

Comment:
why do we need this? we can just change the PartitionInfo's remove method to take a replica id as parameter, and looping the list in PartitionInfo, it's more aligned with addReplicaInfo.

Reply:
right, will change `removeRelicaInfoIfPresent `